radix-engine/src/wasm/wasmi.rs

```diff
@@ -11,279 +14,478 @@ use crate::wasm::constants::*;
+///   - According to the `wasmi` documentation
+///     At the moment `&'a mut System API` which is wrapped into `RadixEngineWasmRuntime` and boxed
+///     is not `Sync`
+///     (the same reason as in alternative `wasmer` implementation)
+///   - Store also shall be `Send + Sync` because of `WasmiModule`, which is cached.
+/// - WasmiInstanceEnv must be clonable (it is a user data for the Store, which is cloned when
+///   WasmiModule is instantiated)
+#[derive(Clone)]
+pub struct WasmiInstanceEnv {
+    runtime_ptr: usize,
 }
 
-pub struct WasmiEnvModule {}
-
-impl ModuleImportResolver for WasmiEnvModule {
-    fn resolve_func(
-        &self,
-        field_name: &str,
-        signature: &wasmi::Signature,
-    ) -> Result<FuncRef, Error> {
-        match field_name {
-            CONSUME_BUFFER_FUNCTION_NAME => Ok(FuncInstance::alloc_host(
-                signature.clone(),
-                CONSUME_BUFFER_FUNCTION_ID,
-            )),
-            INVOKE_METHOD_FUNCTION_NAME => Ok(FuncInstance::alloc_host(
-                signature.clone(),
-                INVOKE_METHOD_FUNCTION_ID,
-            )),
-            INVOKE_FUNCTION_NAME => Ok(FuncInstance::alloc_host(
-                signature.clone(),
-                INVOKE_FUNCTION_ID,
-            )),
-            CREATE_NODE_FUNCTION_NAME => Ok(FuncInstance::alloc_host(
-                signature.clone(),
-                CREATE_NODE_FUNCTION_ID,
-            )),
-            GET_VISIBLE_NODES_FUNCTION_NAME => Ok(FuncInstance::alloc_host(
-                signature.clone(),
-                GET_VISIBLE_NODES_FUNCTION_ID,
-            )),
-            DROP_NODE_FUNCTION_NAME => Ok(FuncInstance::alloc_host(
-                signature.clone(),
-                DROP_NODE_FUNCTION_ID,
-            )),
-            LOCK_SUBSTATE_FUNCTION_NAME => Ok(FuncInstance::alloc_host(
-                signature.clone(),
-                LOCK_SUBSTATE_FUNCTION_ID,
-            )),
-            READ_SUBSTATE_FUNCTION_NAME => Ok(FuncInstance::alloc_host(
-                signature.clone(),
-                READ_SUBSTATE_FUNCTION_ID,
-            )),
-            WRITE_SUBSTATE_FUNCTION_NAME => Ok(FuncInstance::alloc_host(
-                signature.clone(),
-                WRITE_SUBSTATE_FUNCTION_ID,
-            )),
-            UNLOCK_SUBSTATE_FUNCTION_NAME => Ok(FuncInstance::alloc_host(
-                signature.clone(),
-                UNLOCK_SUBSTATE_FUNCTION_ID,
-            )),
-            GET_ACTOR_FUNCTION_NAME => Ok(FuncInstance::alloc_host(
-                signature.clone(),
-                GET_ACTOR_FUNCTION_ID,
-            )),
-            CONSUME_COST_UNITS_FUNCTION_NAME => Ok(FuncInstance::alloc_host(
-                signature.clone(),
-                CONSUME_COST_UNITS_FUNCTION_ID,
-            )),
-            _ => Err(Error::Instantiation(format!(
-                "Function {} not found",
-                field_name
-            ))),
-        }
+impl WasmiInstanceEnv {
```

Comment:
Can we unimplement `Sync` and `Send` for `WasmiInstanceEnv`.

Reply:
I am afraid we cannot since a `Store`  (which keeps`WasmiInstanceEnv` ) shall be `Sync` and `Send` being a part of `WasmiModule` which is cached by `moka::sync::Cache::builder` which requires `Sync` and `Send`.

Reply:
Hmm this is all scaring me too - can we add a detailed comment explaining the hacks we're using:
* https://github.com/paritytech/wasmi/issues/631
* https://github.com/radixdlt/wasmi/commit/84085f82bd4a18e6ecee530e1cb16675c7334df7

Notably it appears that:
* We're duplicating "State" (and with it, instance) from our cache

I'm slightly worried by the reply:
> Yes, as per the Wasm specification the Store is the owner of all instantiated Wasm related state.
An instance only stores references into its associated Store. The Store even owns all Wasm instances associated to it.

If instance contains pointers into `Store`, will cloning `Instance` and `Store` actually work? Or will `Instance` point into the old `Store`? 

Could we have a test which helps check this - I dunno, maybe it EG has a custom blueprint which has a static variable, and saves to that variable - and then calls the same blueprint in a deeper stack frame and check that its static variable has the original state?

Reply:
Actually - I've taken a look, and it's all fine - it doesn't use pointers, it uses a (very simplistic) Arena allocator.

So an `Instance` is just a `Instance(Stored<InstanceIdx>)` which is `Instance(<offset into Store's instance vec>)` which is basically a paired "arena" index into the store and the instance id in that store.

Our cloned stores actually clone everything (all the arenas and the stored id itself) - so they're essentially the same store - but just different data.

So sharing the Store looks fine from a pointer perspective.

And I've also double-checked the wider use of (eg) the runtime_ptr and it looks good.

Reply:
That said, I'd quite like a significantly longer comment on `WasmiModule` explaining why it's fine 👍